pyomo/contrib/solver/base.py

```diff
@@ -0,0 +1,557 @@
+            # default implementation from IntEnum)
+            return format(self.name, format_spec)
+
+        def __str__(self):
+            # Note: Python 3.11 changed the core enums so that the
+            # "mixin" type for standard enums overrides the behavior
+            # specified in __format__.  We will override str() here to
+            # preserve the previous behavior
+            return self.name
+
+    @document_kwargs_from_configdict(CONFIG)
+    @abc.abstractmethod
+    def solve(self, model: _BlockData, **kwargs) -> Results:
+        """
+        Solve a Pyomo model.
+
+        Parameters
+        ----------
+        model: _BlockData
+            The Pyomo model to be solved
+        **kwargs
+            Additional keyword arguments (including solver_options - passthrough
+            options; delivered directly to the solver (with no validation))
+
+        Returns
+        -------
+        results: :class:`Results<pyomo.contrib.solver.results.Results>`
+            A results object
+        """
+
+    @abc.abstractmethod
+    def available(self):
+        """Test if the solver is available on this system.
+
+        Nominally, this will return True if the solver interface is
+        valid and can be used to solve problems and False if it cannot.
+
+        Note that for licensed solvers there are a number of "levels" of
+        available: depending on the license, the solver may be available
+        with limitations on problem size or runtime (e.g., 'demo'
+        vs. 'community' vs. 'full').  In these cases, the solver may
+        return a subclass of enum.IntEnum, with members that resolve to
+        True if the solver is available (possibly with limitations).
+        The Enum may also have multiple members that all resolve to
+        False indicating the reason why the interface is not available
+        (not found, bad license, unsupported version, etc).
+
+        Returns
+        -------
+        available: SolverBase.Availability
+            An enum that indicates "how available" the solver is.
+            Note that the enum can be cast to bool, which will
+            be True if the solver is runable at all and False
+            otherwise.
+        """
+
+    @abc.abstractmethod
+    def version(self) -> Tuple:
+        """
+        Returns
+        -------
+        version: tuple
+            A tuple representing the version
+        """
+
+    def is_persistent(self):
+        """
+        Returns
+        -------
+        is_persistent: bool
+            True if the solver is a persistent solver.
+        """
+        return False
+
+
+class PersistentSolverBase(SolverBase):
+    """
+    Base class upon which persistent solvers can be built. This inherits the
+    methods from the solver base class and adds those methods that are necessary
+    for persistent solvers.
+
+    Example usage can be seen in the Gurobi interface.
+    """
+
+    CONFIG = PersistentSolverConfig()
```

Comment:
I don't disagree; however, the benefit of having this declared is in autodoc generation. Because this is declared, I can use it to generate docs for `PersistentSolverBase` that show the default config items that should be on _all_ persistent solvers.

Reply:
You can use it in the autodoc without storing it on the class:
```python
@document_kwargs_from_configdict(PersistentSolverConfig())
```